src/main/java/com/google/sps/servlets/utils/YouTubeCommentRetriever.java

```diff
@@ -0,0 +1,102 @@
+import com.google.api.client.http.javanet.NetHttpTransport;
+import com.google.api.client.json.JsonFactory;
+import com.google.api.client.json.jackson2.JacksonFactory;
+import com.google.api.services.youtube.YouTube;
+import com.google.api.services.youtube.model.*;
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+import java.util.ArrayList;
+import java.util.List;
+
+/*
+ * Class to retrieve YouTube comments from a designated URL with certain parameters
+ */
+public class YouTubeCommentRetriever {
+  private final JsonFactory JSON_FACTORY = JacksonFactory.getDefaultInstance();
+  // Limit of comments that can be obtained in one request
+  private final Long COMMENT_LIMIT = 100L;
+  // Parameters required by YouTube API to retrieve the comment threads
+  private final String SNIPPET_PARAMETERS = "snippet,replies";
+  private final String ORDER_PARAMETER = "relevance";
+
+  private final String APPLICATION_NAME = "SAY";
+  // TODO: have dev key come from centralized location, rather than being hard-coded.
+  private final String DEVELOPER_KEY = "AIzaSyDYfjWcy1hEe0V7AyaYzgIQm_rT-9XbiGs";
+
+  private YouTube youtubeService = null;
+
+  public YouTubeCommentRetriever() throws Exception {
+    youtubeService = getService();
+  }
+
+  public YouTubeCommentRetriever(YouTube youTube) {
+    youtubeService = youTube;
+  }
+
+  public List<CommentThread> retrieveComments(String url, long maxComments) throws Exception {
+    String nextPageToken = "";
+    long numCommentsLeft = maxComments;
+    long commentQueryLimit = 0;
+    ArrayList allComments = new ArrayList<>();
+    do {
+      // If commentQueryLimit exceeds the number of comments on the video,
+      // The API will simply return all the comments on a video.
+      commentQueryLimit = Math.min(COMMENT_LIMIT, numCommentsLeft);
+      numCommentsLeft -= COMMENT_LIMIT;
+      YouTube.CommentThreads.List commentRequest = generateYouTubeRequest(url, commentQueryLimit);
+      if (nextPageToken != null && nextPageToken != "") {
+        commentRequest.setPageToken(nextPageToken);
+      }
+      CommentThreadListResponse commentResponse = commentRequest.execute();
+      nextPageToken = commentResponse.getNextPageToken();
+      allComments.addAll(commentResponse.getItems());
+    } while (nextPageToken != null && numCommentsLeft > 0);
+    return allComments;
+  }
+
+  /**
+   * @param url video id of the video to have its comments analyzed
+   * @param maxResults is how many comments to be retrieved; Capped out at 100 per request but may
+   *     be reduced for specific queries.
+   * @return A list of comment threads to be aggregated to the overall list.
+   */
+  private YouTube.CommentThreads.List generateYouTubeRequest(String url, long maxResults)
+      throws GeneralSecurityException, IOException {
+    YouTube.CommentThreads.List commentRequest =
+        youtubeService.commentThreads().list(SNIPPET_PARAMETERS);
```

Comment:
I think I found the issue of the null pointer.
You need YouTube mockedYoutube = mock(YouTube.class, RETURNS_DEEP_STUBS);
For more details, you can see the last example on this page: https://site.mockito.org/javadoc/current/org/mockito/Mockito.html#RETURNS_DEEP_STUBS

I talked to Rick about the alternative, and he found out that you can't mock youTube.commentThreads() as it doesn't have a public constructor: http://javadox.com/com.google.apis/google-api-services-youtube/v3-rev142-1.20.0/com/google/api/services/youtube/YouTube.CommentThreads.List.html

So if you do the following it should work:
```
YouTube mockedYoutube = mock(YouTube.class, RETURNS_DEEP_STUBS);
when(youTube.commentThreads().list(Any())).thenReturn(...[your mock results here]);
```